core/domain/translation_domain.py

```diff
@@ -18,9 +18,258 @@
 
 from __future__ import annotations
 
+import copy
+
 from core import utils
 
-from typing import Dict
+from typing import Dict, List
+from typing_extensions import TypedDict
+
+
+TRANSLATABLE_CONTENT_FORMAT_HTML = 'html'
+TRANSLATABLE_CONTENT_FORMAT_UNICODE_STRING = 'unicode'
+TRANSLATABLE_CONTENT_FORMAT_SET_OF_NORMALIZED_STRING = (
+    'set_of_normalized_string')
+TRANSLATABLE_CONTENT_FORMAT_SET_OF_UNICODE_STRING = 'set_of_unicode_string'
+
+
+class TranslatableContentDict(TypedDict):
+    """Dictionary representing TranslatableContent object."""
+
+    content_id: str
+    content: str|List[str]
+    content_type: str
+
+
+class TranslatedContentDict(TypedDict):
+    """Dictionary representing TranslatedContent object."""
+
+    content: str|List[str]
+    needs_update: bool
+
+
+class BaseTranslatableObject:
+    """Base class for all translatable objects."""
+
+    # Private field to track translatable contents in a BaseTranslatableObject.
+    _translatable_contents: Dict[str, TranslatableContent] = {}
+
+    def _register_all_translatable_fields(self) -> None:
+        """Method to register all translatable fields in a translatable object.
+
+        Raises:
+            NotImplementedError. The derived child class must implement the
+                necessary logic to register all translatable fields in an
+                entity.
+        """
+        raise NotImplementedError('Must be implemented in subclasses.')
+
+    def _register_translatable_field(
+        self,
+        field_type: str,
+        content_id: str,
+        value: str|List[str]
```

Comment:
Here and elsewhere -- should we make a custom type for this, since we use it a lot and it's kind of it's own "conceptual thing" at this point?

Reply:
Done